### ISSUE TITLE
Question : Stage 3 LM preparation too slow

### ISSUE THREAD

**Issue by dsa934** [espnet/espnet#1100]
Currently I am using gtx1660ti (single gpu)

I am running the librispeech recipe, 
but after 4096 minutes it does not proceed to the next step(stage 4).

What should I do?

The run.sh file says that for a single gpu, change the accum_grad and batch-bin. 
Is there an exact figure?

**Comment by sw005320** [comment]
> I am running the librispeech recipe, but after 4096 minutes it does not proceed to the next step(stage 4).

Can you tell me more exactly what happened?

> The run.sh file says that for a single gpu, change the accum_grad and batch-bin. Is there an exact figure?

It depends on the batch size, but in my experiment, `accum-grad: 8` gave us reasonable performance with a single GPU.

**Comment by dsa934** [comment]
Computer Spec : ubuntu 18.04LTS, GTX 1660ti ( single GPU), 64GB RAM, i7-9700k @ 3.6GHz(CPU)

When I first ran espnet / egs / librispeech / asr1 / run.sh, I got a cuda out of memory error.
so I changed ngpu in run.sh file from 4 to 1 and ran run.sh again. 
then Stage 0 ~ 2 went without problems. 

However,
the Stage_3 (LM preparation) phase has not progressed to the next stage, Stage_4, for 4 days.
If i check with nvidia-smi, top, jobs, the process is still in progress.
in your espnet's issue, I heard that using librispeech with a single gpu takes about four days.

But in my case, it's been more than 4 days in LM preparation
Is it originally going to take such a long time or do I need to change more settings and proceed?

I ran the program by running the run.sh file and looking at the log file and modifying it
in this case, it just takes a long time without any error.
there was no action I could take right now, so I left a question.

**Comment by Fhrozen** [comment]
You may reference to #572 for librispeech trainin in single gpu. The specs of the card where GTX 1080 Ti for 3.5 days. Now, you can check some specs on both cards like [here](https://www.gpucheck.com/compare/nvidia-geforce-gtx-1080-ti-vs-nvidia-geforce-gtx-1660-ti/intel-core-i7-7700k-4-20ghz-vs-intel-core-i7-8700k-3-70ghz/).

The GTX 1660 ti  is a Turing processor with better clock speeds that the 1080 but it has only 6Gb.
The main difference will be on the Memory bandwidth where is only has 288Gbps (1080> 320Gbps) and while the GTX 1660 has better performance for fp16 ( 10,000 Gflops) for fp32  only has 5k Gflops (while 1080 has 8k Gflops). 

In your case it may be take one or two more days that the already established. 
If you are on a hurry, please download the already trained model described at: https://github.com/espnet/espnet/blob/master/egs/librispeech/asr1/RESULTS.md

**Comment by sw005320** [comment]
> You may reference to #572 for librispeech trainin in single gpu. The specs of the card where GTX 1080 Ti for 3.5 days. Now, you can check some specs on both cards like [here](https://www.gpucheck.com/compare/nvidia-geforce-gtx-1080-ti-vs-nvidia-geforce-gtx-1660-ti/intel-core-i7-7700k-4-20ghz-vs-intel-core-i7-8700k-3-70ghz/).

We changed the LM config from #527 setup. Now, we used very large LSTMLM as default, and it took a very long time now (more than a week). 

@dsa934, it depends on your purpose, but if you just want to get some reasonable performance, you can just use the old language model (https://github.com/espnet/espnet/blob/master/egs/librispeech/asr1/conf/tuning/lm.yaml). It would take a few days. 

We also now support the multi-GPU training for LM (#1083). You can test this as well although it may change the performance.

**Comment by dsa934** [comment]
Thank you for reply.

In summary, if I use LargeLM with my computer, it will take a week or more in Stage_3 only.
and it takes longer to proceed with the whole recipe?

So you told me that I had to use the old LM model to run the recipe faster.
all I have to do is modify run.sh code?
( e.g  lm_config = conf/lm.yaml  -> lm_config=conf/tuning/lm.yaml )


And if this recipe works in the future, can I plug in my custom model and run it?

**Comment by sw005320** [comment]
> In summary, if I use LargeLM with my computer, it will take a week or more in Stage_3 only.
and it takes longer to proceed with the whole recipe?

Yes.

> So you told me that I had to use the old LM model to run the recipe faster.
all I have to do is modify run.sh code?
( e.g lm_config = conf/lm.yaml -> lm_config=conf/tuning/lm.yaml )

Correct.

> And if this recipe works in the future, can I plug in my custom model and run it?

It depends on how you customize your model, but it should be.

**Comment by dsa934** [comment]
Sorry, There is one more question.
I ran run.sh using the old LM. ( e.g lm_config = conf/lm.yaml -> lm_config=conf/tuning/lm.yaml )
and then 

( In stage 3)
processed 40560000 lines
processed 40570000 lines
...
processed 40690000 lines
skipped 1 empty lines
filtered 0 lines
skipped 0 empty lines
filtered 0 lines

After this message, the process is in progress without a message. 
Is it normal to take long on this part?

When I first used large LSTM_LM it took a long time in this part. So I ask you this question.
